### DISCUSSION TITLE
What's the best way to secure the endpoints?

### DISCUSSION THREAD

**Discussion by pbasov** [substratusai/kubeai#217] [Q&A]
For LLMs I understand I can put everything behind OpenWebUI and let users manage their keys there, but for faster-whisper, it doesn't seem like owui proxies /audio/transcriptions calls through their endpoint. 

I looked into a bunch of solutions like Kong and Istio, but they all seem needlessly complicated, especially when you want users to have self-service.

Relevant issue I guess: https://github.com/substratusai/kubeai/issues/101
But i think it doesn't have to be built into the endpoints, a reverse-proxy is fine.

**Comment by pbasov** [comment]
One solution is to have LiteLLM, but they have paywalled the SSO feature unfortunately.

**Reply by samos123** [reply]
What did you end up using? Something like Kong is an option too which would allow you to do things like rate limiting. 

Kong also seems to support SSO: https://docs.konghq.com/konnect/api/identity-management/sso/

Also apologies for missing this for so long. We have mostly been focusing on issues and not on the Discussions part.

**Reply by pbasov** [reply]
The easiest solution was https://gateway.envoyproxy.io/latest/tasks/security/apikey-auth/

To create these secrets I've created a small controller with a frontend.

**Reply by nstogner** [reply]
Envoy proxy via Gateway API makes a lot of sense, especially if you want consistent auth across multiple components where KubeAI is one. +1 to this approach.

**Reply by nstogner** [reply]
PS: We might include a built-in feature in the future that ties an api key to token usage / quota. If you think that would be useful, let us know.

**Reply by pbasov** [reply]
It's definitely something that customers want. But with envoy your API key is tied to an ingress and not to the vllm instance where you get the token count. How would you count the tokens on the ingress side?
